File: 18.4/ex4.py
from abc import ABC, abstractmethod

class Employee(ABC):
    def __init__(self, name):
        self.name = name

    @abstractmethod
    def earnings(self):
        pass

class Boss(Employee):
    def __init__(self, name, monthly_salary):
        super().__init__(name)
        self.monthly_salary = monthly_salary

    def earnings(self):
        return self.monthly_salary

class PieceWorker(Employee):
    def __init__(self, name, num_products):
        super().__init__(name)
        self.num_products = num_products

    def earnings(self):
        return self.num_products * 0.5

class CommissionWorker(Employee):
    def __init__(self, name, base_salary, num_products_sold):
        super().__init__(name)
        self.base_salary = base_salary
        self.num_products_sold = num_products_sold

    def earnings(self):
        return self.base_salary + (self.num_products_sold * 0.1)

def main():
    boss_name = input("Enter boss name: ")
    boss_salary = float(input("Enter boss monthly salary: "))
    boss = Boss(boss_name, boss_salary)
    print(f"{boss.name} earns: ${boss.earnings()}")

    piece_worker_name = input("Enter piece worker name: ")
    num_products = int(input("Enter number of products for piece worker: "))
    piece_worker = PieceWorker(piece_worker_name, num_products)
    print(f"{piece_worker.name} earns: ${piece_worker.earnings()}")

    commission_worker_name = input("Enter commission worker name: ")
    base_salary = float(input("Enter commission worker base salary: "))
    num_products_sold = int(input("Enter number of products sold by commission worker: "))
    commission_worker = CommissionWorker(commission_worker_name, base_salary, num_products_sold)
    print(f"{commission_worker.name} earns: ${commission_worker.earnings()}")

if __name__ == "__main__":
    main()

File: 18.4/ex4.java
import java.util.Scanner;

abstract class Employee {
    protected String name;

    public Employee(String name) {
        this.name = name;
    }

    public abstract double earnings();
}

class Boss extends Employee {
    private double monthlySalary;

    public Boss(String name, double monthlySalary) {
        super(name);
        this.monthlySalary = monthlySalary;
    }

    public double earnings() {
        return this.monthlySalary;
    }
}

class PieceWorker extends Employee {
    private int numProducts;

    public PieceWorker(String name, int numProducts) {
        super(name);
        this.numProducts = numProducts;
    }

    public double earnings() {
        return this.numProducts * 0.5;
    }
}

class CommissionWorker extends Employee {
    private double baseSalary;
    private int numProductsSold;

    public CommissionWorker(String name, double baseSalary, int numProductsSold) {
        super(name);
        this.baseSalary = baseSalary;
        this.numProductsSold = numProductsSold;
    }

    public double earnings() {
        return this.baseSalary + (this.numProductsSold * 0.1);
    }
}

public class ex4 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter boss name: ");
        String bossName = scanner.nextLine();
        System.out.print("Enter boss monthly salary: ");
        double bossSalary = scanner.nextDouble();
        Boss boss = new Boss(bossName, bossSalary);
        System.out.println(boss.name + " earns: $" + boss.earnings());

        scanner.nextLine(); 

        System.out.print("Enter piece worker name: ");
        String pieceWorkerName = scanner.nextLine();
        System.out.print("Enter number of products for piece worker: ");
        int numProducts = scanner.nextInt();
        PieceWorker pieceWorker = new PieceWorker(pieceWorkerName, numProducts);
        System.out.println(pieceWorker.name + " earns: $" + pieceWorker.earnings());

        scanner.nextLine(); 

        System.out.print("Enter commission worker name: ");
        String commissionWorkerName = scanner.nextLine();
        System.out.print("Enter commission worker base salary: ");
        double baseSalary = scanner.nextDouble();
        System.out.print("Enter number of products sold by commission worker: ");
        int numProductsSold = scanner.nextInt();
        CommissionWorker commissionWorker = new CommissionWorker(commissionWorkerName, baseSalary, numProductsSold);
        System.out.println(commissionWorker.name + " earns: $" + commissionWorker.earnings());

        scanner.close();
    }
}
